### ISSUE TITLE
Unable to recreate DNM backprojection using MNIST data

### ISSUE THREAD

**Issue by emilykaczmarek** [mpslxz/DNM#1]
Hello,

I have been trying to run the DNM code with the MNIST data, but have been unable to recreate the backprojection image. I ran the code with the same parameters as in the Deep Neural Maps paper, with 500 pretraining iterations for the autoencoder, 10000 for the SOM, and 200 for the DNM, and the following image is the final recreated backprojection: 

![DNM_test_dnm_training](https://user-images.githubusercontent.com/52331761/64045091-739f4100-cb36-11e9-8970-1f1e45c7ebfe.png)

I saved various backprojections throughout the pretraining of the SOM, and the following image never changed until the DNM training, regardless of what iteration it was on: 

![DNM_test_som_training](https://user-images.githubusercontent.com/52331761/64045097-78fc8b80-cb36-11e9-8391-bed9b02a7e07.png)

When I run the code for 5 iterations each of ae, SOM, and DNM with the initial parameters as the DNM_intermediate.npy checkpoints saved from the previous trial (of 500, 10000, and 200 iterations), I get the following backprojection:

![DNM_test_intermediate](https://user-images.githubusercontent.com/52331761/64045318-f1fbe300-cb36-11e9-8e28-936b1b8cd02e.png)

However, if I use the DNM.npy checkpoints from the 500-10000-200 training as the initial parameters, I get this:

![DNM_test_dnm_ckpts](https://user-images.githubusercontent.com/52331761/64045358-093ad080-cb37-11e9-835e-7ee725a97620.png)

I'm trying to get the third image directly when I run the code the first time. I have been trouble shooting for some time now and am unable to fix this problem. Any help would be much appreciated!

**Comment by mpslxz** [comment]
Can you check the reconstructions via DNM.npy?

**Comment by emilykaczmarek** [comment]
The last photo I posted is the reconstruction from DNM.npy, sorry for the confusion!

**Comment by mpslxz** [comment]
That is the map, I believe. I wanted to see whether or not your AE is reconstructing an image correctly.

**Comment by emilykaczmarek** [comment]
The image below is one of the reconstructions using DNM.npy. It looks like the AE is reconstructing the images properly. 
![reconstructed_test](https://user-images.githubusercontent.com/52331761/64705054-49764900-d47d-11e9-9977-8f5c3c89d99a.png)

**Comment by mpslxz** [comment]
Could you post your training log?

**Comment by mpslxz** [comment]
BTW, I ran the code again and got the following map:

![DNM_projection](https://user-images.githubusercontent.com/7834847/64978856-51622e80-d86b-11e9-8846-4b47456f5eb2.png)

I think you can increase the learning rate (dnm_map_lr) to 0.5 to reach the fine tuning faster.

**Comment by emilykaczmarek** [comment]
Here is the training log for 500 iterations of the ae, 1500 iterations of the SOM (with learning rate 0.005), and 200 iterations of the DNM. I can also run the full 10000 iterations of the SOM with the original learning rate of 0.0005 if that's helpful. 

https://docs.google.com/document/d/10e2XsbuVjDkPIG7GmshxWqszNSHIfAzDPt_6XiwcPxI/edit?usp=sharing

**Comment by mpslxz** [comment]
This is strange. Your error log says that the scan function is not available. Essentially, scan is what the model is using to update the SOM weights for all of the samples in a batch. Also, I am getting different distances than yours, when I run the code. One quick thing for you to try is increasing the dnm training epochs from 200 to 1000 or so and the dnm learning rate to 0.5 or so. Seems that, for whatever reason during joint optimization, SOM updates are slower than what it should be.

**Comment by emilykaczmarek** [comment]
I changed the DNM learning rate to 0.5 and the number of epochs to 1000, and this is what I got for the backprojection: 

![DNM_1000iterations](https://user-images.githubusercontent.com/52331761/65433906-e75a0400-ddeb-11e9-9241-7a11f468bdd2.png)

Then, when I ran the code with the saved intermediate and final weights, I got these two backprojections:

![DNM_intermediate](https://user-images.githubusercontent.com/52331761/65433952-f9d43d80-ddeb-11e9-8e88-1654df53e47a.png)

![DNM_final](https://user-images.githubusercontent.com/52331761/65433956-fc369780-ddeb-11e9-85ce-6a58ce868a26.png)

It seems like the weights of the SOM are saving in the checkpoints, but for some reason aren't saving in the actual code - maybe this is because scan isn't available?

**Comment by mpslxz** [comment]
Which checkpoint did you load and continue training from (the last two figures)?

**Comment by emilykaczmarek** [comment]
The first figure used the intermediate checkpoint (DNM_intermediate.npy), and the second figure used the last checkpoint (DNM.npy). I didn't do any further training once I loaded the checkpoints, just set the dnm_epochs and pre_train_epochs to 0.

**Comment by mpslxz** [comment]
Seems that the problem has been solved. I am closing this issue.
